File: src/picalc/picalc.ml
let xorshift32 state =
  let x = !state in
  let x = Int32.logxor x (Int32.shift_left x 13) in
  let x = Int32.logxor x (Int32.shift_right_logical x 17) in
  let x = Int32.logxor x (Int32.shift_left x 5) in
  state := x;
  let unsigned =
    Int64.logand (Int64.of_int32 x) 0xFFFF_FFFFL
  in
  (Int64.to_float unsigned) /. 4294967295.0

let worker start_seed iterations =
  let state = ref (Int32.of_int start_seed) in
  let inside = ref 0 in
  for _ = 1 to iterations do
    let x = xorshift32 state in
    let y = xorshift32 state in
    if x *. x +. y *. y <= 1.0 then incr inside
  done;
  !inside

let () =
  let total_iters = if Array.length Sys.argv > 1 then int_of_string Sys.argv.(1) else 50_000_000 in
  let num_threads =
    if Array.length Sys.argv > 2 then max 1 (int_of_string Sys.argv.(2)) else 4
  in
  
  let domains = Array.init num_threads (fun i ->
    let start = (total_iters * i) / num_threads in
    let finish = (total_iters * (i + 1)) / num_threads in
    Domain.spawn (fun () -> worker (12345 + i) (finish - start))
  ) in
  
  let total_inside = Array.fold_left (fun acc d -> acc + Domain.join d) 0 domains in
  let pi = 4.0 *. float_of_int total_inside /. float_of_int total_iters in
  Printf.printf "%.5f\n" pi

File: src/picalc/picalc.nim
import std/[os, strutils, threadpool]

proc xorshift32(state: var uint32): float64 =
  var x = state
  x = x xor (x shl 13)
  x = x xor (x shr 17)
  x = x xor (x shl 5)
  state = x
  return float64(x) / 4294967295.0

proc worker(start_seed: uint32, iterations: int): int =
  var state = start_seed
  var inside = 0
  for i in 0..<iterations:
    let x = xorshift32(state)
    let y = xorshift32(state)
    if x * x + y * y <= 1.0:
      inside.inc
  return inside

proc main() =
  let totalIters = if paramCount() > 0: parseInt(paramStr(1)) else: 50_000_000
  let numThreads = if paramCount() > 1: max(1, parseInt(paramStr(2))) else: 4
  
  var responses: seq[FlowVar[int]]
  for i in 0..<numThreads:
    let start = (totalIters * i) div numThreads
    let finish = (totalIters * (i + 1)) div numThreads
    responses.add(spawn worker(12345'u32 + uint32(i), finish - start))
    
  var totalInside = 0
  for i in 0..<numThreads:
    totalInside += ^responses[i]
    
  let pi = 4.0 * float64(totalInside) / float64(totalIters)
  echo formatFloat(pi, ffDecimal, 5)

main()
